quipucords/scanner/network/utils.py

```diff
@@ -173,3 +181,54 @@ def get_fact_names():
 def raw_facts_template():
     """Results template for fact collection on network scans."""
     return {fact_name: None for fact_name in get_fact_names()}
+
+
+GEN_KEYFILES_DIRECTORY_PREFIX = "quipucords_ssh_private_key_files_"
+GEN_KEYFILE_PREFIX = "ssh_private_key_file_credential_"
+
+
+def generate_ssh_keyfile(credential, ssh_keyvalue):
+    """Generate an ssh private keyfile for the ssh key content specified."""
+    private_keyfile_path = os.path.join(
+        tempfile.mkdtemp(prefix=GEN_KEYFILES_DIRECTORY_PREFIX),
+        f"{GEN_KEYFILE_PREFIX}{credential.get('id')}",
+    )
+    private_key_fp = open(private_keyfile_path, "w+")
+    private_key_fp.write(ssh_keyvalue)
+    private_key_fp.write("\n")
+    private_key_fp.close()
+    os.chmod(private_keyfile_path, 0o600)
```

Comment:
Mentioned in another comment, but adding here for local context. We can probably simplify and harden this code by using [tempfile.TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) and [tempfile.NamedTemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile).

Reply:
To clarify and expand on what I mean here, I think _each file_ we write to disk should have its own `NamedTemporaryFile` inside its own `TemporaryDirectory`.  We _could_ have them all share one `TemporaryDirectory`, but directories are cheap. We would need to retain in local memory a reference to the `NamedTemporaryFile` and `TemporaryDirectory` that we would use later for deletion.

Using `tempfile` objects like that would significantly reduce the risk of multiple threads/processes stomping on each others' files, and it means _we_ aren't responsible for choosing arbitrary path names. We can completely drop `GEN_KEYFILES_DIRECTORY_PREFIX` and `GEN_KEYFILE_PREFIX` since the names of these files don't really matter as long as we can pass the name to Ansible. Python's library is already smart enough to create _very_ unique names that will almost never conflict. It will actually test the name it generates, and it will keep trying a new random names if the last name it generated already exists.

Reply:
I think that's fine.  Other than the deletion upon context exit which we don't use, TemporaryDirectory is the same as (and actually uses) mkdtemp. for the filename itself, I had used the {GEN_KEYFILE_PREFIX}{credential.get('id')} pattern thinking of not creating thousands of files if our scans of 1K or 10K system use the same credential/ssh key (most likely).  Then again, if mkdtemp() is random, it's a moot point.  Hmm, Maybe we do need a unique directory for this process/thread/scan. thoughts ?

Reply:
If we're ok with thousands of SSH key files, we could stick them all in one directory (not random named) but with random filenames, at least they're not all over the place and the cleanup code will be simpler (no need to delete the parent directory), kind'a like a qpc .ssh directory but for dynamically generated files.

Reply:
> Other than the deletion upon context exit which we don't use, TemporaryDirectory is the same

I think having the automatic cleanup on exit would still be good free extra protection in case we fail to delete any individual file. I just tested this locally to confirm that automatic cleanup does not actually require using `TemporaryDirectory` as a context manager. It looks like cleanup also happens on garbage collection on normal process shutdown:

```
>>> import tempfile
>>> d = tempfile.TemporaryDirectory()
>>> d.name
'/var/folders/71/5v1_8dbd03j_nfxbb8bkb1q00000gn/T/tmpq2zufxkw'
>>> from pathlib import Path
>>> Path(f'{d.name}/test.txt').touch()
```

Meanwhile in a separate shell while that REPL is still open, just to be extra sure:

```
❯ ls /var/folders/71/5v1_8dbd03j_nfxbb8bkb1q00000gn/T/tmpq2zufxkw
test.txt
```

and back to the first REPL:

```
>>> glob.glob(f"{d.name}/*")
['/var/folders/71/5v1_8dbd03j_nfxbb8bkb1q00000gn/T/tmpq2zufxkw/test.txt']
>>> ^d

❯ ls /var/folders/71/5v1_8dbd03j_nfxbb8bkb1q00000gn/T/tmpq2zufxkw
ls: /var/folders/71/5v1_8dbd03j_nfxbb8bkb1q00000gn/T/tmpq2zufxkw: No such file or directory
```

It's gone! 🎉 

`mkdtemp` does not do similar cleanup on exit:

```
>>> import tempfile
>>> d = tempfile.mkdtemp()
>>> d
'/var/folders/71/5v1_8dbd03j_nfxbb8bkb1q00000gn/T/tmpicmk21_c'
>>> from pathlib import Path
>>> Path(f'{d}/test.txt').touch()
>>> import glob
>>> glob.glob(f'{d}/*')
['/var/folders/71/5v1_8dbd03j_nfxbb8bkb1q00000gn/T/tmpicmk21_c/test.txt']
>>> ^d

❯ ls /var/folders/71/5v1_8dbd03j_nfxbb8bkb1q00000gn/T/tmpicmk21_c
test.txt
```

It's still there! 💥 

> then I used the {GEN_KEYFILE_PREFIX}{credential.get('id')} pattern thinking of not creating thousands of files

In a simpler case, I wouldn't oppose using a regular pattern like `{GEN_KEYFILE_PREFIX}{credential.get('id')}`, but with our scanning process, the risk here is that if you have two scans running in parallel using the same credentials but for maybe different sources, you run the risk of one deleting the file before the other is finished using it. So, we would have to get "clever" and add some randomness, at which point we might simply use full random name generation, thus `NamedTemporaryFile`.

I think having thousands of files in a directory should not be a problem here. As long as we're cleaning up after ourselves, those files should be relatively short-lived anyway, and I _doubt_ (but woudl love to be proven wrong) many of our current users actually have thousands of credentials or thousands of sources (not to be confused with thousands of hosts in a source) given how difficult that would be to define in the GUI or CLI. If I read our code correctly, and please correct me if I'm wrong, this new temporary file is written to disk effectively once per source, not once per host, which is good for keeping the file count low.

* once per source+credential combination:
  * connect task -> construct_inventory -> _construct_vars -> _credential_vars -> generate_ssh_keyfile
  * inspect task -> construct_inventory -> _construct_vars -> _credential_vars -> generate_ssh_keyfile

Is my reading here correct?

On a general note, I've worked on filesystems with upwards of 10k files in one directory without any major problems. Any problems I've seen typically arise from having too many _open_ file handles at once, but I don't think we're in that space here yet.

Reply:
That's what I thought so too (one per source), but it does create one per target host. So for now, I will update the code to create them all in one directory (no need to create/delete thousands of directories) but will move to NamedTemporaryFile for guaranteeing uniqueness and additional protection (their deletion upon process termination in case something got missed (error/exception)). 

I will push an update shortly.
Thank you.

Reply:
https://github.com/quipucords/quipucords/pull/2390/commits/48899b1ba6715b9de413e3e3c267225a1a70aae4

Reply:
I _think_ the recent changes have worked around this risk, if I'm reading the diff correctly. Thanks!